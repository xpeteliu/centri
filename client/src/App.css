.mainContent {
  margin: 20px;
}

body {
  background-color: #E4F0D1;
}

.btn-primary, .btn-primary:hover, .btn-primary:active, .btn-primary:visited {
  background-color: #96BB88 !important;
  border-color: #96BB88 !important;
  color: #000000 !important;
}

.btn-secondary, .btn-secondary:hover, .btn-secondary:active, .btn-secondary:visited {
  background-color: #C1D9B8 !important;
  border-color: #C1D9B8 !important;
  color: #000000 !important;
}

.headerBar {
  background-color: #C1D9B8 !important;
}

.groupListItem {
  background-color: #FFFFFF;
  height: 80px;
  padding: 20px;
}

/* .App {
  text-align: center;
}

.App-logo {
  height: 40vmin;
  pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-float infinite 3s ease-in-out;
  }
}

.App-header {
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
}

.App-link {
  color: rgb(112, 76, 182);
}

@keyframes App-logo-float {
  0% {
    transform: translateY(0);
  }
  50% {
    transform: translateY(10px);
  }
  100% {
    transform: translateY(0px);
  }
} */
