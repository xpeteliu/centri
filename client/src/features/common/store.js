import { configureStore } from '@reduxjs/toolkit';
import modalReducer from './MessageModal/modalSlice';

const store = configureStore({
  reducer: {
    modal: modalReducer,
  },
});

export default store;
