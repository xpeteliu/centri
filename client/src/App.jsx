import React, { Component } from 'react';
import {
  BrowserRouter as Router,
  Route
} from 'react-router-dom';
import HomePage from './components/HomePage/HomePage';
import LoginPage from './components/LoginPage/LoginPage';
import './App.css';

const Home = () => (
  <HomePage />
);

const Login = () => (
  <LoginPage />
);

class App extends Component {
  render() {
    return (
      <Router>
        <div className="App">
          <Route exact path="/" component={Home} />
          <Route path="/login" component={Login} />
        </div>
      </Router>
    );
  }
}

export default App;
