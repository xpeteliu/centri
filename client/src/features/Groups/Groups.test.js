import React from 'react';
import {
  act, render, screen,
} from '@testing-library/react';
import { Router } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import { createMemoryHistory } from 'history';
import { Provider } from 'react-redux';
import { setupServer } from 'msw/node';
import { rest } from 'msw';
import mockServer from '../common/mockServer.test';
import App from '../../App';
import store from '../common/store';
import { configureStore } from '@reduxjs/toolkit';
import userReducer from '../common/userSlice';
import modalReducer from '../common/MessageModal/modalSlice';

describe('Group List page', () => {
  const testGroup = {
    _id: '123',
    title: 'Test Group',
    tags: [],
    status: 'public',
    adminIds: ['1234'],
    memberIds: ['12345'],
    pendingMemberIds: ['123'],
  };

  const mockServerGroupList = setupServer(
    rest.post('/api/group/filter/sort',
      (req, res, ctx) => res(ctx.json([testGroup]))),
    rest.get('/api/group/123',
      (req, res, ctx) => res(ctx.json(testGroup))
    )
  );

  beforeAll(() => { mockServerGroupList.listen(); });
  afterAll(() => { mockServerGroupList.close(); });

  const history = createMemoryHistory({ initialEntries: ['/groups'] });

  test('should match snapshot', async () => {
    let container;
    act(() => {
      ({ container } = render(
        <Provider store={store}>
          <Router history={history}>
            <App />
          </Router>
        </Provider>,
      ));
    });
    expect(container).toMatchSnapshot();
  });

  test('renders groups, not yet joined', async () => {
    const userStore = configureStore({
      reducer: {
        modal: modalReducer,
        user: userReducer,
      },
      preloadedState: { user: { id: '123456' } },
    });
    act(() => {
      render(
        <Provider store={userStore}>
          <Router history={history}>
            <App />
          </Router>
        </Provider>,
      );
    });
    expect(await screen.findByText('Test Group')).toBeVisible();
    const joinButton = await screen.findByRole('button', { name: /Join Group/ });
    expect(joinButton).toBeVisible();
    act(() => {
      userEvent.click(joinButton);
    });
    const requestedButton = await screen.findByRole('button', { name: /Requested/ });
    expect(requestedButton).toBeDisabled();
  });

  test('renders groups, already joined', async () => {
    const userStore = configureStore({
      reducer: {
        modal: modalReducer,
        user: userReducer,
      },
      preloadedState: { user: { id: '1234' } },
    });
    act(() => {
      render(
        <Provider store={userStore}>
          <Router history={history}>
            <App />
          </Router>
        </Provider>,
      );
    });
    expect(await screen.findByText('Test Group')).toBeVisible();
    const requestedButton = await screen.findByRole('button', { name: /Requested/ });
    expect(requestedButton).toBeDisabled();
  });

  test('creates group', async () => {
    act(() => {
      render(
        <Provider store={store}>
          <Router history={history}>
            <App />
          </Router>
        </Provider>,
      );
    });
    const createBtn = await screen.findByRole('button', { name: /Create Group/ });
    act(() => {
      userEvent.click(createBtn);
    });
    const inputName = await screen.getByPlaceholderText('Enter group name');
    act(() => {
      userEvent.type(inputName, 'Group Name 4');
    });
    const createBtn2 = await screen.findAllByRole('button', { name: /Create Group/ });
    act(() => {
      userEvent.click(createBtn2[1]);
    });
  });

  test('group buttons', async () => {
    act(() => {
      render(
        <Provider store={store}>
          <Router history={history}>
            <App />
          </Router>
        </Provider>,
      );
    });
    const createBtn = await screen.findByRole('button', { name: /Create Group/ });
    expect(createBtn).toBeVisible();
    const myGroups = await screen.findByRole('button', { name: /My Groups/ });
    expect(myGroups).toBeVisible();
    const publicGroups = await screen.findByRole('button', { name: /My Groups/ });
    expect(publicGroups).toBeVisible();
  });
});

describe('Group/Admin pages', () => {
  const testGroup = {
    _id: '123',
    title: 'Test Group',
    tags: [],
    status: 'public',
    adminIds: ['1234'],
    memberIds: ['12345'],
    pendingMemberIds: ['123'],
  };
  const testPost = {
    _id: '1234',
    groupId: '123',
    heading: 'Test Post',
    creatorId: '1234',
    content: 'Test Content',
    flaggerId: '12345',
  };
  const mockServerGroup = setupServer(
    rest.post('/api/group/filter/sort',
      (req, res, ctx) => res(ctx.json([testGroup]))),
    rest.get('/api/group/123',
      (req, res, ctx) => res(ctx.json(testGroup))
    ),
    rest.post('/api/posting/filter/paginate',
      (req, res, ctx) => res(ctx.json([testPost]))
    ),
    rest.get('/api/posting/1234',
      (req, res, ctx) => res(ctx.json(testPost))
    ),
  );
  beforeAll(() => { mockServerGroup.listen(); });
  afterAll(() => { mockServerGroup.close(); });

  test('should match snapshot', async () => {
    let container;
    const history = createMemoryHistory({ initialEntries: ['/group/123'] });
    act(() => {
      ({ container } = render(
        <Provider store={store}>
          <Router history={history}>
            <App />
          </Router>
        </Provider>,
      ));
    });
    expect(container).toMatchSnapshot();
  });

  test('group page as admin', async () => {
    const userStore = configureStore({
      reducer: {
        modal: modalReducer,
        user: userReducer,
      },
      preloadedState: { user: { id: '1234' } },
    });
    const history = createMemoryHistory({ initialEntries: ['/group/123'] });
    act(() => {
      render(
        <Provider store={userStore}>
          <Router history={history}>
            <App />
          </Router>
        </Provider>,
      );
    });
    expect(await screen.findByText('Test Group')).toBeVisible();
    const adminButton = await screen.findByRole('button', { name: /Admin/ });
    expect(adminButton).toBeVisible();
    act(() => {
      userEvent.click(adminButton);
    });
  });

  test('group page as non-admin', async () => {
    const userStore = configureStore({
      reducer: {
        modal: modalReducer,
        user: userReducer,
      },
      preloadedState: { user: { id: '12345' } },
    });
    const history = createMemoryHistory({ initialEntries: ['/group/123'] });
    act(() => {
      render(
        <Provider store={userStore}>
          <Router history={history}>
            <App />
          </Router>
        </Provider>,
      );
    });
    expect(await screen.findByText('Test Group')).toBeVisible();
    expect(await screen.findByText('Test Post')).toBeVisible();
    expect(await screen.findByText('Test Content')).toBeVisible();
    const editButton = await screen.findByRole('button', { name: /Edit Group/ });
    expect(editButton).toBeVisible();
    const analyticsButton = await screen.findByRole('button', { name: /Analytics/ });
    expect(analyticsButton).toBeVisible();
    const createPostButton = await screen.findByRole('button', { name: /Create Post/ });
    expect(createPostButton).toBeVisible();
    const leaveButton = await screen.findByRole('button', { name: /Leave Group/ });
    expect(leaveButton).toBeVisible();
  });

  test('admin page', async () => {
    const userStore = configureStore({
      reducer: {
        modal: modalReducer,
        user: userReducer,
      },
      preloadedState: { user: { id: '1234' } },
    });
    const history = createMemoryHistory({ initialEntries: ['/group/123/admin'] });
    act(() => {
      render(
        <Provider store={userStore}>
          <Router history={history}>
            <App />
          </Router>
        </Provider>,
      );
    });
    expect(await screen.findByText('Pending Requests')).toBeVisible();
    expect(await screen.findByText('Admins')).toBeVisible();
    expect(await screen.findByText('Members')).toBeVisible();
    expect(await screen.findByText('Flagged Posts')).toBeVisible();
    const demoteButton = await screen.findByRole('button', { name: /Demote/ });
    expect(demoteButton).toBeVisible();
    const deleteFlaggedButton = await screen.findByRole('button', { name: /Delete/ });
    expect(deleteFlaggedButton).toBeVisible();
    const backButton = await screen.findByRole('button', { name: /Back to Group/ });
    expect(backButton).toBeVisible();
    // expect(await screen.findByText('Pending Requests')).toBeVisible();
  });
});
