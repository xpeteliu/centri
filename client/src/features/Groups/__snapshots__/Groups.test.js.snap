// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Group page should match snapshot 1`] = `
<div>
  <div
    class="App"
  >
    <div
      class="App container"
    >
      <nav
        class="headerBar navbar navbar-expand-md navbar-light bg-light"
      >
        <div
          class="container"
        >
          <a
            class="navbar-brand"
            href="/"
          >
            Logo
          </a>
          <button
            aria-label="Toggle navigation"
            class="navbar-toggler collapsed"
            type="button"
          >
            <span
              class="navbar-toggler-icon"
            />
          </button>
          <div
            class="navbar-collapse collapse"
          >
            <div
              class="me-auto navbar-nav"
            >
              <a
                class="nav-link"
                data-rr-ui-event-key="/groups"
                href="/groups"
              >
                Groups
              </a>
              <a
                class="nav-link"
                data-rr-ui-event-key="/messages"
                href="/messages"
              >
                Messages
              </a>
            </div>
            <div
              class="hstack gap-2"
            >
              <img
                class="rounded-circle"
                height="30"
                src="./photo.jpg"
                width="30"
              />
              <span
                class="navbar-text"
              >
                Justin Dong
              </span>
            </div>
          </div>
        </div>
      </nav>
      <div
        class="mainContent vstack gap-5"
      >
        <div
          class="groupHeader row"
        >
          <div
            class="groupListItem hstack gap-4"
          >
            <div
              class="bg-light"
            >
              <img
                class="rounded-circle"
                height="30"
                src="./photo.jpg"
                width="30"
              />
            </div>
            <div>
              <h3>
                Group Name
              </h3>
            </div>
            <div
              class="ms-auto"
            >
              <button
                class="btn btn-primary"
                type="button"
              >
                Edit Group
              </button>
            </div>
          </div>
        </div>
        <div
          class="groupList row"
        >
          <div
            class="vstack gap-5"
          >
            <div
              class="card"
            >
              <div
                class="card-body"
              >
                <div
                  class="card-title h5"
                >
                  Post Title (id:
                  0
                  )
                </div>
                <p
                  class="card-text"
                >
                  Vitae massa id tortor sed tortor, commodo. Platea blandit mauris elementum est maecenas iaculis. Ut sodales diam, nam commodo gravida faucibus nisl. Aliquet id pulvinar id lacinia. Lectus auctor vel elementum tristique. Vitae non morbi dolor quisque amet faucibus justo. Sollicitudin vitae augue tortor lobortis sem ultrices neque. Eget ornare et varius in lectus. Lectus est ante morbi ipsum.         Aliquam sit non viverra suspendisse eleifend. Et mauris, et quam dolor duis. Lacus porttitor felis, a, vel sed enim. Quam nisi, est nulla scelerisque sollicitudin faucibus erat. Tincidunt purus mauris felis fringilla sit. Commodo dignissim amet vel in ultrices sagittis ultrices. Sem malesuada donec nam eget a, risus laoreet. Egestas malesuada ipsum lacus, quis. A eleifend dolor, id tincidunt diam tincidunt. Tellus suspendisse ut luctus mauris bibendum. Eu sed in convallis neque. In amet convallis sit eros, leo. Id volutpat sit morbi sagittis neque.         Dolor, tortor aliquet dictumst mattis mi, netus in. Egestas blandit nunc nulla eget in lacus a, sit. Nulla.
                </p>
              </div>
            </div>
            <div
              class="card"
            >
              <div
                class="card-body"
              >
                <div
                  class="card-title h5"
                >
                  Post Title (id:
                  1
                  )
                </div>
                <p
                  class="card-text"
                >
                  Vitae massa id tortor sed tortor, commodo. Platea blandit mauris elementum est maecenas iaculis. Ut sodales diam, nam commodo gravida faucibus nisl. Aliquet id pulvinar id lacinia. Lectus auctor vel elementum tristique. Vitae non morbi dolor quisque amet faucibus justo. Sollicitudin vitae augue tortor lobortis sem ultrices neque. Eget ornare et varius in lectus. Lectus est ante morbi ipsum.         Aliquam sit non viverra suspendisse eleifend. Et mauris, et quam dolor duis. Lacus porttitor felis, a, vel sed enim. Quam nisi, est nulla scelerisque sollicitudin faucibus erat. Tincidunt purus mauris felis fringilla sit. Commodo dignissim amet vel in ultrices sagittis ultrices. Sem malesuada donec nam eget a, risus laoreet. Egestas malesuada ipsum lacus, quis. A eleifend dolor, id tincidunt diam tincidunt. Tellus suspendisse ut luctus mauris bibendum. Eu sed in convallis neque. In amet convallis sit eros, leo. Id volutpat sit morbi sagittis neque.         Dolor, tortor aliquet dictumst mattis mi, netus in. Egestas blandit nunc nulla eget in lacus a, sit. Nulla.
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Groups page should match snapshot 1`] = `
<div>
  <div
    class="App"
  >
    <div
      class="App container"
    >
      <nav
        class="headerBar navbar navbar-expand-md navbar-light bg-light"
      >
        <div
          class="container"
        >
          <a
            class="navbar-brand"
            href="/"
          >
            Logo
          </a>
          <button
            aria-label="Toggle navigation"
            class="navbar-toggler collapsed"
            type="button"
          >
            <span
              class="navbar-toggler-icon"
            />
          </button>
          <div
            class="navbar-collapse collapse"
          >
            <div
              class="me-auto navbar-nav"
            >
              <a
                class="nav-link"
                data-rr-ui-event-key="/groups"
                href="/groups"
              >
                Groups
              </a>
              <a
                class="nav-link"
                data-rr-ui-event-key="/messages"
                href="/messages"
              >
                Messages
              </a>
            </div>
            <div
              class="hstack gap-2"
            >
              <img
                class="rounded-circle"
                height="30"
                src="./photo.jpg"
                width="30"
              />
              <span
                class="navbar-text"
              >
                Justin Dong
              </span>
            </div>
          </div>
        </div>
      </nav>
      <div
        class="mainContent vstack gap-5"
      >
        <div
          class="groupButtons row"
        >
          <div
            class="col-4"
          >
            <button
              class="btn btn-secondary"
              type="button"
            >
              Private Groups
            </button>
          </div>
          <div
            class="col-4"
          >
            <button
              class="btn btn-secondary"
              type="button"
            >
              Public Groups
            </button>
          </div>
          <div
            class="col-4"
          >
            <button
              class="btn btn-primary"
              type="button"
            >
              Create Group
            </button>
          </div>
        </div>
        <div
          class="groupList row"
        >
          <div
            class="vstack gap-5"
          >
            <div
              class="groupListItem hstack gap-4"
            >
              <div
                class="bg-light"
              >
                <img
                  class="rounded-circle"
                  height="30"
                  src="./photo.jpg"
                  width="30"
                />
              </div>
              <div>
                <h3>
                  <a
                    href="./group"
                  >
                    Group Name
                  </a>
                </h3>
              </div>
              <div
                class="ms-auto"
              >
                <button
                  class="btn btn-secondary"
                  type="button"
                >
                  Join Group
                </button>
              </div>
            </div>
            <div
              class="groupListItem hstack gap-4"
            >
              <div
                class="bg-light"
              >
                <img
                  class="rounded-circle"
                  height="30"
                  src="./photo.jpg"
                  width="30"
                />
              </div>
              <div>
                <h3>
                  <a
                    href="./group"
                  >
                    Group Name 2
                  </a>
                </h3>
              </div>
              <div
                class="ms-auto"
              >
                <button
                  class="btn btn-secondary"
                  type="button"
                >
                  Join Group
                </button>
              </div>
            </div>
            <div
              class="groupListItem hstack gap-4"
            >
              <div
                class="bg-light"
              >
                <img
                  class="rounded-circle"
                  height="30"
                  src="./photo.jpg"
                  width="30"
                />
              </div>
              <div>
                <h3>
                  <a
                    href="./group"
                  >
                    Group Name 3
                  </a>
                </h3>
              </div>
              <div
                class="ms-auto"
              >
                <button
                  class="btn btn-secondary"
                  type="button"
                >
                  Join Group
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
